File: DailyChallenges/Scala/LeetCode_785.scala
object Solution {
    def isBipartite(graph: Array[Array[Int]]): Boolean = {
        val n = graph.length 
        val coloured = Array.fill[Int](n)(0) // array of ball pending colour
        
        // helper function to color the ball
        def colour(a: Array[Int], c: Int): Unit = for (i <- a if coloured(i) == 0) {
            coloured(i) = c
            colour(graph(i), -c)
        }
        
        // colour the balls
        for {i <- Range(0, n) if coloured(i) == 0} {
            coloured(i) = 1
            colour(graph(i), -1)
        }
        
        // check if isBipartite
        graph.zipWithIndex.forall(p => p._1.forall(coloured(_) == -coloured(p._2)))
    }
}

File: DailyChallenges/CPP/LeetCode_785.cpp
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        vector<int> ans(graph.size(), 0); // bw => 1 or -1 
        function<bool(int, int)> mark = [&](int i, int bw){
            if (ans[i] == bw) return true;
            else if (ans[i] == -bw) return false;
            else {
                ans[i] = bw;
                for(int j: graph[i])
                    if (!mark(j, -bw))
                        return false;
                return true;
            }
        };
        for(int i=0; i<graph.size(); ++i){
            if (ans[i]) continue;
            if (!mark(i, 1)) return false;
        }
        return true;
    }
};
